Add NextDNS integration

bieniu opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->


## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
This PR adds a new integration [NextDNS](https://nextdns.com).

This PR is the first step, I plan to add `diagnostics`, `system_health`, `binary_sensor`, `button` and `switch` platforms in the next PRs.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/23234

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:
re
- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

bieniu committed:
Initial commit

bieniu committed:
Update manifest

bieniu committed:
Add first test

bieniu committed:
Simplify init

bieniu committed:
More tests

bieniu committed:
Update tests

bieniu committed:
More tests

bieniu committed:
More tests

bieniu committed:
Add tests for sensor platform

bieniu committed:
More tests for sensor platform

bieniu committed:
Add tests for system_health

bieniu committed:
Fix typo

bieniu committed:
Improve test coverage

bieniu committed:
Improve test coverage

bieniu committed:
Add tests for diagnostics

bieniu committed:
Add comment

bieniu committed:
Run hassfest

bieniu committed:
Fix typo

bieniu committed:
Run gen_requirements_all

bieniu committed:
Fix tests

bieniu committed:
Change key name in diagnostics

allenporter commented:
My impression is the best practice is to add a single platform in the initial integration PR. I appreciate that others are saved for follow up PRs, though i think it makes sense to pick one platform for the initial review based on past guidance i've seen.

bieniu committed:
Remove diagnostics and system_health platforms

bieniu commented:
@allenporter `diagnostics` and `system_health` platforms have been removed.

allenporter commented on homeassistant/components/nextdns/config_flow.py:
```diff
@@ -0,0 +1,88 @@
+                errors["base"] = "invalid_api_key"
+            except (ApiError, ClientConnectorError, asyncio.TimeoutError):
+                errors["base"] = "cannot_connect"
+            except Exception:  # pylint: disable=broad-except
```
Omit this and only catch specific known exceptions

bieniu replied:
What if there is an exception that we can't predict right now? I believe this is a good protection against unknown errors.

allenporter replied:
It's against the design standards for the code base to catch raw Exception, isn't it? The problem is that it mixes programming errors with other types of errors.

bieniu replied:
IMO sometimes we should skip standards when the situation requires it. Exactly 130 other integrations catches a general exception in config flow.
Mayby we should second opinion here 😀

gjohansson-ST replied:
I think for a config flow it makes sense to catch broad exception to provide feedback to the user and not simply break the flow.

allenporter replied:
I'm not seeing anything here that requires this (it seems arbitrary for this call.. any other line of could could also have a bug and throw an unexpected exception too) so happy to have someone else review and approve. Looks great otherwise.

allenporter replied:
I wasn't aware this is part of the scaffold script template, which is helpful context.

https://github.com/home-assistant/core/blob/dev/script/scaffold/templates/config_flow/integration/config_flow.py#L92

allenporter commented on homeassistant/components/nextdns/config_flow.py:
```diff
@@ -0,0 +1,88 @@
+            data_schema=vol.Schema(
+                {
+                    vol.Required(CONF_PROFILE_NAME): vol.In(
+                        [profile.name for profile in self.nextdns.profiles]
```
Do you need to handle the case where there are no existing profiles to choose from?  I'm assuming you have to ask the user to set one up. Or is there always a profile?

bieniu replied:
Each time a new NextDNS account is created, there is one default profile available, we don't need to ask the user to create a profile.

allenporter commented on homeassistant/components/nextdns/sensor.py:
```diff
@@ -0,0 +1,299 @@
+        icon="mdi:dns",
+        name="{profile_name} DNS Queries",
+        native_unit_of_measurement="queries",
+        state_class=SensorStateClass.MEASUREMENT,
```
Since this is a # of queries, i'm wondering if this is really `TOTAL` or `TOTAL_INCREASING` since it's the total count of queries, rather than a measurement of something like "queries per second"

Caveat: I am not a sensor expert.

bieniu replied:
`TOTAL_INCREASING` does't fit here as the statistics can be reset and the value will drop to zero.

According to the documentation `TOTAL` state class "should not be used for sensors where the absolute value is interesting instead of the accumulated growth or decline". I think the absolute value is interesting here.

allenporter replied:
Ah, OK, I had in mind that only the rate (QPS) is interesting, but i'm used to measuring queries in another context. Thanks for confirming.

allenporter commented on homeassistant/components/nextdns/sensor.py:
```diff
@@ -0,0 +1,299 @@
+        super().__init__(coordinator)
+        self._attr_device_info = coordinator.device_info
+        self._attr_unique_id = f"{coordinator.profile_id}_{description.key}"
+        self._attr_name = cast(str, description.name).format(
```
I couldn't find where `name` is defined to understand why the cast is needed. Can you point me to where the name is defined in the entity description?

bieniu replied:
`name` is defined here https://github.com/home-assistant/core/blob/dev/homeassistant/helpers/entity.py#L225 and the type of it is `str | None`. `mypy` complains that `Item "None" of "Optional[str]" has no attribute "format"`.

allenporter reviewed (changes requested):
This looks like a very clean implementation, great work. 👍🏼  I had a few smaller comments and questions, but overall this is in great shape.

allenporter commented:
May need rebase after https://github.com/home-assistant/core/pull/74329 for tests to pass? Not sure

bieniu committed:
Bump library

allenporter commented:
Oh forgot: Do you have a link to the changelog for the library?

bieniu commented:
@allenporter Of course https://github.com/bieniu/nextdns/compare/0.1.0...1.0.0
I added unit tests, added one new feature that will be used with switch platform and fixed one small bug.

Thanks for the review.

gjohansson-ST commented on homeassistant/components/nextdns/sensor.py:
```diff
@@ -0,0 +1,299 @@
+    @callback
+    def _handle_coordinator_update(self) -> None:
+        """Handle updated data from the coordinator."""
+        self._attr_native_value = getattr(
```
Could use a lambda in description instead

bieniu replied:
I think I need help here, what exactly do you mean?

gjohansson-ST replied:
I mean something like this https://github.com/home-assistant/core/blob/dev/homeassistant/components/sensibo/sensor.py#L241

bieniu replied:
Smart, I will address all your comments in the new PR.

gjohansson-ST commented on homeassistant/components/nextdns/sensor.py:
```diff
@@ -0,0 +1,299 @@
+    NextDnsSensorEntityDescription(
+        key="all_queries",
+        coordinator_type=ATTR_STATUS,
+        entity_category=EntityCategory.DIAGNOSTIC,
```
Are all of these really for diag?

bieniu replied:
In my opinion, yes, this is information about the number of queries with different criteria. Do you think another entity category should be used?

gjohansson-ST replied:
I don't know. Just asking since it appears a bit strange all sensors goes as diag and not any off them are without a category.

gjohansson-ST commented on homeassistant/components/nextdns/__init__.py:
```diff
@@ -0,0 +1,185 @@
+
+    tasks = []
+
+    # Independent DataUpdateCoordinator is used for each API endpoint to avoid
+    # unnecessary requests when entities using this endpoint are disabled.
```
Wouldn't they register and start updating anyhow or am I missing something?

bieniu replied:
All coordinators are registered and update data during HA startup. After that, only coordinators with enabled entities will update the data.

gjohansson-ST replied:
Really. Need to remember that.

gjohansson-ST reviewed (commented):
Added some comments

bieniu commented on homeassistant/components/nextdns/__init__.py:
```diff
@@ -0,0 +1,185 @@
+    except (ApiError, ClientConnectorError, asyncio.TimeoutError) as err:
+        raise ConfigEntryNotReady from err
+
+    hass.data.setdefault(DOMAIN, {})
+    hass.data[DOMAIN].setdefault(entry.entry_id, {})
```
I think I don't understand. Dou tou mean `hass.data.setdefault(DOMAIN, {})[entry.entry_id] = {}`?

gjohansson-ST replied:
Yes. I just meant you can combine them into one line

bieniu replied:
Yup, good idea.

bieniu commented on homeassistant/components/nextdns/__init__.py:
```diff
@@ -0,0 +1,185 @@
+            hass, nextdns, profile_id, update_interval
+        )
+        tasks.append(
+            hass.data[DOMAIN][entry.entry_id][coordinator_name].async_refresh()
```
Yes, `async_config_entry_first_refresh()` should be used here. Good point.

bieniu commented on tests/components/nextdns/__init__.py:
```diff
@@ -0,0 +1,63 @@
+        data={CONF_API_KEY: "fake_api_key", CONF_PROFILE_ID: "xyz12"},
+    )
+
+    if not add_to_hass:
```
I'll change this.

bieniu commented on homeassistant/components/nextdns/config_flow.py:
```diff
@@ -0,0 +1,88 @@
+
+    def __init__(self) -> None:
+        """Initialize the config flow."""
+        self.nextdns: NextDns
```
@MartinHjelmare Done https://github.com/home-assistant/core/pull/75635
